File: LeetCode/P215_KthLargestElement.java
/*
 * LeetCode 215 - Kth Largest Element in an Array
 * https://leetcode.com/problems/kth-largest-element-in-an-array/
 *
 * Approach: Min-Heap of size k
 * Time  : O(n log k)
 * Space : O(k)
 */

import java.util.PriorityQueue;

public class P215_KthLargestElement {

    public int findKthLargest(int[] nums, int k) {
        PriorityQueue<Integer> minHeap = new PriorityQueue<>();

        for (int num : nums) {
            minHeap.offer(num);
            if (minHeap.size() > k) {
                minHeap.poll();
            }
        }
        return minHeap.peek();
    }

    // ---------- Simple manual test ----------
    public static void main(String[] args) {
        P215_KthLargestElement sol = new P215_KthLargestElement();

        int[] nums1 = {3, 2, 1, 5, 6, 4};
        System.out.println(sol.findKthLargest(nums1, 2)); // 5

        int[] nums2 = {3, 2, 3, 1, 2, 4, 5, 5, 6};
        System.out.println(sol.findKthLargest(nums2, 4)); // 4
    }
}

File: LeetCode/P215_KthLargestElement.cpp
#include <bits/stdc++.h>
using namespace std;

class P215_KthLargestElement {
public:
    int findKthLargest(vector<int>& nums, int k) {
        priority_queue<int, vector<int>, greater<int>> minHeap;

        for (int num : nums) {
            minHeap.push(num);
            if ((int) minHeap.size() > k) {
                minHeap.pop();
            }
        }
        return minHeap.top();
    }
};

int main() {
    P215_KthLargestElement sol;

    vector<int> nums1 = {3, 2, 1, 5, 6, 4};
    cout << sol.findKthLargest(nums1, 2) << "\n";

    vector<int> nums2 = {3, 2, 3, 1, 2, 4, 5, 5, 6};
    cout << sol.findKthLargest(nums2, 4) << "\n";

    return 0;
}
